samcli/commands/logs/logs_context.py

```diff
@@ -119,22 +122,67 @@ def _fetch_resources_from_stack(self, selected_resource_names: Optional[Set[str]
 
         Returns
         -------
-        List[StackResourceSummary]
+        List[CloudFormationResourceSummary]
             List of resource information, which will be used to fetch the logs
         """
-        results = []
         LOG.debug("Getting logical id of the all resources for stack '%s'", self._stack_name)
         stack_resources = get_resource_summaries(
             self._boto_resource_provider, self._stack_name, ResourcePhysicalIdResolver.DEFAULT_SUPPORTED_RESOURCES
         )
 
-        if selected_resource_names is None:
-            selected_resource_names = {stack_resource.logical_resource_id for stack_resource in stack_resources}
+        if selected_resource_names:
+            return self._get_selected_resources(stack_resources, selected_resource_names)
+        return self._get_all_resources(stack_resources)
+
+    def _get_selected_resources(
+        self,
+        resource_summaries: List[CloudFormationResourceSummary],
+        selected_resource_names: Set[str],
+    ) -> List[CloudFormationResourceSummary]:
+        """
+        Returns list of resources which matches with selected_resource_names.
+        selected_resource_names can be;
+        - resource name like HelloWorldFunction
+        - or it could be pointing to a resource in nested stack like NestedApp/HelloWorldFunction
 
-        for resource in stack_resources:
-            # if resource name is not selected, continue
-            if resource.logical_resource_id not in selected_resource_names:
-                LOG.debug("Resource (%s) is not selected with given input", resource.logical_resource_id)
-                continue
-            results.append(resource)
-        return results
+        Parameters
+        ----------
+        resource_summaries : List[CloudFormationResourceSummary]
+            List of CloudformationResourceSummary which was returned from given stack
+        selected_resource_names : Set[str]
+            List of resource name definitions that will be used to filter the results
+
+        Returns
+        ------
+        List[CloudFormationResourceSummary]
+            Filtered list of CloudFormationResourceSummary's
+        """
+        resources = []
+        for selected_resource_name in selected_resource_names:
```

Comment:
It will iterate the tree for each selected_resource_name, use a dictionary (resources_id -> resource_summary) may improve the performance.